### ISSUE TITLE
Creating Signing and Submiting Multi Sig Transaction With Nami Wallet

### ISSUE THREAD

**Issue by edridudi** [bloxbean/cardano-client-lib#44]
Below code demonstrates MintTransaction Creation at backend -> Passing it to Nami wallet at the frontend -> Nami User signs the transaction (partially) and sends it back to the backend for Policy Signing and Wallet Account Signing.

this code works only for One Asset mint and it isn't working for multiple Assets as a bundle.

The issue I encounter is:
`Result{successful=false, response='{"error":"Bad Request","message":"\"transaction submit error ShelleyTxValidationError ShelleyBasedEraAlonzo (ApplyTxError [UtxowFailure (WrappedShelleyEraFailure (InvalidWitnessesUTXOW [VKey (VerKeyEd25519DSIGN \\\"f09dd265566b154afd7082a3a0d7daaa210c64ba286fefc43bcff398b07e23ea\\\")]))])\"","status_code":400}', code=400, value=null}`

What am I missing?

### Step One ->
Transaction Creation which is then serialized and sent to Nami wallet at the frontend.

```
public Transaction createTransaction(MintRequest mintRequest, Policy policy, String senderAddress) throws ApiException, CborSerializationException, CborDeserializationException {
        //For Mint
        RequireTimeBefore requireTimeBefore = null;
        for (NativeScript nativeScript : policy.getScriptAll().getScripts()) {
            if (nativeScript instanceof RequireTimeBefore) {
                requireTimeBefore = (RequireTimeBefore) nativeScript;
            }
        }
        long ttl = getBlockService().getLastestBlock().getValue().getSlot() + 2000;
        if (requireTimeBefore != null) {
            ttl = requireTimeBefore.getSlot();
        }
        TransactionDetailsParams detailsParams = TransactionDetailsParams.builder().ttl(ttl).build(); 

        List<MultiAsset> multiAssetList = Collections.singletonList(mintRequest.getMultiAsset());

        BigInteger minAmount = createDummyOutputAndCalculateMinAdaForTxnOutput(senderAddress, multiAssetList, protocolParams);
        BigInteger amountToTransfer = BigInteger.valueOf(mintRequest.getCost());

        //Build Raw Transaction

        // Fetch utxos from backend
        List<Utxo> utxos = getUtxos(senderAddress, LOVELACE, amountToTransfer.add(ONE_ADA.multiply(BigInteger.valueOf(2))), utxosToExclude);
//        List<Utxo> utxos2 = utxoSelectionStrategy.selectUtxos(senderAddress, LOVELACE, amountToTransfer.add(ONE_ADA.multiply(BigInteger.valueOf(2))), Collections.EMPTY_SET); //transfer amount + 2 ADA to cover fee and min ada

        if (utxos.isEmpty())
            throw new InsufficientBalanceException("Not enough utxos found to cover balance : " + amountToTransfer + " lovelace");

        //Create single TxnOutput for the sender
        TransactionOutput change = TransactionOutput
                .builder()
                .address(senderAddress)
                .value(Value.builder().coin(BigInteger.ZERO)
                        .multiAssets(new ArrayList<>())
                        .build())
                .build();

        //Inputs
        List<TransactionInput> inputs = new ArrayList<>();
        for (Utxo utxo: utxos) {
            TransactionInput input = TransactionInput.builder()
                    .transactionId(utxo.getTxHash())
                    .index(utxo.getOutputIndex()).build();
            inputs.add(input);

            copyUtxoValuesToChangeOutput(change, utxo);
        }

        //Outputs
        TransactionOutput output = TransactionOutput
                .builder()
                .address(mintRequest.getAddress())
                .value(new Value(amountToTransfer, null))
                .build();


        //Deduct fee + minCost in a MA output
        BigInteger remainingAmount = change.getValue().getCoin().subtract(amountToTransfer.add(minAmount));
        change.getValue().setCoin(remainingAmount); //deduct requirement amt (fee + min amount)

        //Check if minimum Ada is not met. Topup
        //Transaction will fail if minimum ada not there. So try to get some additional utxos
        verifyMinAdaInOutputAndUpdateIfRequired(inputs, change, detailsParams, utxos, protocolParams);

        //Create a separate output for minted assets
        //Create output
        TransactionOutput mintedTransactionOutput = new TransactionOutput();
        mintedTransactionOutput.setAddress(senderAddress);
        Value value = Value.builder()
                .coin(minAmount)
                .multiAssets(new ArrayList<>())
                .build();
        mintedTransactionOutput.setValue(value);
        for (MultiAsset ma : multiAssetList) {
            mintedTransactionOutput.getValue().getMultiAssets().add(ma);
        }
        List<TransactionOutput> outputs = Arrays.asList(output, change, mintedTransactionOutput);

        TransactionBody body = TransactionBody.builder()
                .inputs(inputs)
                .outputs(outputs)
                .fee(BigInteger.valueOf(170000))
                .ttl(detailsParams.getTtl())
                .validityStartInterval(detailsParams.getValidityStartInterval())
                .mint(multiAssetList)
                .build();

        // Sign ->

        TransactionWitnessSet transactionWitnessSet = new TransactionWitnessSet();
        transactionWitnessSet.getNativeScripts().add(policy.getScriptAll());

        Transaction transaction = Transaction.builder()
                .body(body)
                .witnessSet(transactionWitnessSet)
                .metadata(mintRequest.getMetaData())
                .build();

        Account signer1;
        Account signer2;

        if (configProperties.getNetwork() == Network.TEST_NET) {
            //Signers
            String signerAcc1Mnemonic = "around submit turtle canvas friend remind push vehicle debate drop blouse piece obvious crane tone avoid aspect power milk eye brand cradle tide wrist";
            String signerAcc2Mnemonic = "prison glide olympic diamond rib payment crucial ski vintage example dinner matrix cruise upper antenna surge drink divorce brother half figure skate jar stand";
            signer1 = new Account(Networks.testnet(), signerAcc1Mnemonic);
            signer2 = new Account(Networks.testnet(), signerAcc2Mnemonic);
        }

        //Sign the transaction. so that we get the actual size of the transaction to calculate the fee
        String signTxnHash = signer1.sign(transaction); //cbor encoded bytes in Hex format
        signTxnHash = signer2.sign(signTxnHash);

        if (policy.getPolicyKeys()!= null) {
            signTxnHash = CardanoJNAUtil.signWithSecretKey(signTxnHash, HexUtil.encodeHexString(policy.getPolicyKeys().getSkey().getBytes()));
        }

        BigInteger fee = feeCalculationService.calculateFee(Transaction.deserialize(HexUtil.decodeHexString(signTxnHash)));

        body.setFee(fee);
        //Final change amount after amountToTransfer + fee
        change.getValue().setCoin(change.getValue().getCoin().subtract(fee));
        return transaction;
    } 
```

### Step Two ->
**txnHash** Represents Nami Returned witnessSet hash after user sign.
**transaction** represents the original transaction from previous step.
```
public String signAndSubmit(int id, String txnHash, Transaction transaction) throws CborSerializationException, ApiException, CborException, CborDeserializationException {
        if (transaction==null) {
            return null;
        }

        //Nami Sign Deserialize
        TransactionWitnessSet transactionWitnessSet = TransactionWitnessSet.deserialize((Map) CborDecoder.decode(HexUtil.decodeHexString(txnHash)).get(0));
        transaction.getWitnessSet().getVkeyWitnesses().add(transactionWitnessSet.getVkeyWitnesses().get(0));

        // Policy Sign
        Policy policy = policyService.findPolicy(pendingConcurrentHashMap.get(id).getPolicyId());
        String signer1TxnHash = CardanoJNAUtil.signWithSecretKey(transaction.serializeToHex(), HexUtil.encodeHexString(policy.getPolicyKeys().getSkey().getBytes()));

        // Account Sign
        signer1TxnHash = walletService.getAccountByWalletIndex(id).sign(Transaction.deserialize(HexUtil.decodeHexString(signer1TxnHash)));

        Result<String> result = cardanoApiService.getTransactionService().submitTransaction(HexUtil.decodeHexString(signer1TxnHash));

        log.info(result.toString());
        if (result.isSuccessful()) {
            return result.getValue();
        } else {
            return null;
        }
    }
```

**Comment by satran004** [comment]
@edridudi I have not gone in detail. But just a quick observation. 

It could be due to the issue mentioned in CIP21 (https://cips.cardano.org/cips/cip21/)  HW compatiblity. In multiasset section it says
```Since multiassets (policy_id and asset_name) are represented as maps, both need to be sorted in accordance with the specified canonical CBOR format. ```

In cardano-client-lib, this CIP is not yet implemented. But most probably it would be part of  the next release.

Can you please check if the multiassets order changes after you de-serialize a Nami signed transaction? I am suspecting, Nami is already implementing CIP21.

**Comment by edridudi** [comment]
@satran004 
Nami wallet returns witnessSet in cbor hex format.
It is then deserialized and its vkey content is added to the original transaction.

**Comment by satran004** [comment]
Thanks @edridudi  Is it possible to sign with same mnemonic (which Nami is using) in Java and check if the signature is same in  witnessset ? I am suspecting it would be different.. if Nami is already sorting multple assets (as mentioned in CIP21) as witness signature is done for hash(transactionBody).

If that's the case, one workaround could be :- try the  sorting of multiasset list in your Java code as mentioned in the CIP 

I will also try it later and let you know.

**Comment by satran004** [comment]
@edridudi, Just to make sure multiasset ordering is an issue here, another workaround you may want to try for now.

1. Instead of signing first with Nami, do policy sign in Java. This will  go through de-serialization and serialization in native lib which may change the order of the multiasset list as required.

2.  Get the Transaction object  by deserializing the policy signed transaction hex.
3. Clear the witnessmap.
4. Serialize and pass the updated transaction/transaction body to Nami for signing without policy witness. (I am not sure if Nami needs the whole transaction or  transaction body is enough)
5. Add Nami returned witness to the transaction object.
6. If the policy witness is not stored in the backend, do policy sign again in  Java.

**Comment by satran004** [comment]
It seems "nonCanonical()" encoding in Transaction.serialize() needs to be removed to fix this issue. Though I need to verify this.  @edridudi you may want to compile cardano-client-lib and try.

https://github.com/bloxbean/cardano-client-lib/blob/560266e27557ddb6576501d40086fe339ca7302a/src/main/java/com/bloxbean/cardano/client/transaction/spec/Transaction.java#L102

After change, it should be something like this
```
new CborEncoder(baos).encode(cborBuilder.build());
```

**Comment by edridudi** [comment]
@satran004 
removed "nonCanonical()" encoding from Transaction.serialize() for version 0.1.5

Error is now:

`Result{successful=false, response='{"error":"Bad Request","message":"\"transaction submit error ShelleyTxValidationError ShelleyBasedEraAlonzo (ApplyTxError [UtxowFailure (WrappedShelleyEraFailure (InvalidWitnessesUTXOW [VKey (VerKeyEd25519DSIGN \\\"f09dd265566b154afd7082a3a0d7daaa210c64ba286fefc43bcff398b07e23ea\\\")])),UtxowFailure (WrappedShelleyEraFailure (ConflictingMetadataHash (AuxiliaryDataHash {unsafeAuxiliaryDataHash = SafeHash \\\"a4d942eabbafebef4933db7aba4563b58611b1f0eedb2cd620cab8c8945af258\\\"}) (AuxiliaryDataHash {unsafeAuxiliaryDataHash = SafeHash \\\"6375c2f84b2c8d336df2e8b5e23d180563613a0a60de6882a5195f83ef709100\\\"})))])\"","status_code":400}', code=400, value=null}`

It seems like I need to use the latest code base.
I will do that today.

**Comment by satran004** [comment]
Thanks @edridudi . If removing nonCanonical is not solving it, you may face the same issue with the latest code.
I am  trying to reproduce it with your sample code. I will update my findings later.

**Comment by satran004** [comment]
Can you please also give me  your test token names in mint transaction ?
 I just tried with canonical encoding with my Nami account and it somehow worked. 
 Here's the txn hash https://testnet.cardanoscan.io/transaction/13a25294ce7653b0458e39e901ddc40b448432fcb21f9939bed1bb3f42a5ee25?tab=tokenmint

**Comment by edridudi** [comment]
that's awesome !
maybe it works because you are using latest code base?

anyway I'm still working on the 0.1.5 Release and I noticed that the Original Transaction is changed after Policy sign.

before:
`TransactionBody(inputs=[TransactionInput{transactionId=fd917babf04a421d00568b6ee69c481b1669b9cb528fa4c31171046e8e3fb50e, index=1}], outputs=[TransactionOutput{address='addr_test1qrcl5nghjs98eldm5fzymmlrk2qnk2a8x8pmuglcuf2r7s7tzpxm2u4mva7ywq0xssx6a03mgeu82jns03g6ln0w7npszr2lpt', value=Value(coin=90000000, multiAssets=null)}, TransactionOutput{address='addr_test1qz0xcyfuwkf6a2c8g0mhjdaxxvtuw2u04dqjx7tt2gwaq5522z65y7wauh6rryspdn7xrg5u7nkf5ung6qk5dn3a7u8syvce7n', value=Value(coin=871618084, multiAssets=[])}, TransactionOutput{address='addr_test1qz0xcyfuwkf6a2c8g0mhjdaxxvtuw2u04dqjx7tt2gwaq5522z65y7wauh6rryspdn7xrg5u7nkf5ung6qk5dn3a7u8syvce7n', value=Value(coin=1551690, multiAssets=[MultiAsset{policyId=654ebfc69ea9b582d09755a0760fdac7b3e16718ef47acd958708035, assets=[Asset{name=WeedBox255, value=1}, Asset{name=MusicBong380, value=1}, Asset{name=WeedBox265, value=1}]}])}], fee=244173, ttl=62345848, mint=[MultiAsset{policyId=654ebfc69ea9b582d09755a0760fdac7b3e16718ef47acd958708035, assets=[Asset{name=WeedBox255, value=1}, Asset{name=MusicBong380, value=1}, Asset{name=WeedBox265, value=1}]}], metadataHash=[1, 87, -41, -13, 63, 73, 102, 19, -84, -126, -15, -69, 5, -39, 106, 29, 88, 84, -6, 59, 20, 109, 80, -92, 17, 10, 100, -89, -84, -17, 44, 22], validityStartInterval=0)`

After:
`Transaction(body=TransactionBody(inputs=[TransactionInput{transactionId=fd917babf04a421d00568b6ee69c481b1669b9cb528fa4c31171046e8e3fb50e, index=1}], outputs=[TransactionOutput{address='addr_test1qrcl5nghjs98eldm5fzymmlrk2qnk2a8x8pmuglcuf2r7s7tzpxm2u4mva7ywq0xssx6a03mgeu82jns03g6ln0w7npszr2lpt', value=Value(coin=90000000, multiAssets=[])}, TransactionOutput{address='addr_test1qz0xcyfuwkf6a2c8g0mhjdaxxvtuw2u04dqjx7tt2gwaq5522z65y7wauh6rryspdn7xrg5u7nkf5ung6qk5dn3a7u8syvce7n', value=Value(coin=871618084, multiAssets=[])}, TransactionOutput{address='addr_test1qz0xcyfuwkf6a2c8g0mhjdaxxvtuw2u04dqjx7tt2gwaq5522z65y7wauh6rryspdn7xrg5u7nkf5ung6qk5dn3a7u8syvce7n', value=Value(coin=1551690, multiAssets=[MultiAsset{policyId=654ebfc69ea9b582d09755a0760fdac7b3e16718ef47acd958708035, assets=[Asset{name=0x4d75736963426f6e67333830, value=1}, Asset{name=0x57656564426f78323535, value=1}, Asset{name=0x57656564426f78323635, value=1}]}])}], fee=244173, ttl=62345848, mint=[MultiAsset{policyId=654ebfc69ea9b582d09755a0760fdac7b3e16718ef47acd958708035, assets=[Asset{name=0x4d75736963426f6e67333830, value=1}, Asset{name=0x57656564426f78323535, value=1}, Asset{name=0x57656564426f78323635, value=1}]}], metadataHash=[1, 87, -41, -13, 63, 73, 102, 19, -84, -126, -15, -69, 5, -39, 106, 29, 88, 84, -6, 59, 20, 109, 80, -92, 17, 10, 100, -89, -84, -17, 44, 22], validityStartInterval=0), witnessSet=TransactionWitnessSet(vkeyWitnesses=[VkeyWitness(vkey=[-16, -99, -46, 101, 86, 107, 21, 74, -3, 112, -126, -93, -96, -41, -38, -86, 33, 12, 100, -70, 40, 111, -17, -60, 59, -49, -13, -104, -80, 126, 35, -22], signature=[94, -122, -38, 14, -47, 82, -44, 92, 93, -23, 109, -114, 79, 37, 75, 26, -54, 34, 58, 62, -83, -107, -109, -17, -23, -27, 54, -86, 1, 102, -1, 104, -128, 58, 117, -21, -72, -25, 15, -80, 120, 64, 25, 61, 4, -56, -57, 16, -57, -5, -72, -99, 76, -127, -101, -103, -117, -93, -79, -12, -37, -57, 65, 8]), VkeyWitness(vkey=[83, 96, -2, 13, -118, 64, -10, 110, 28, -18, 72, -96, 84, 99, -63, 69, 28, -33, -25, -91, 111, 3, 67, 73, -72, 74, -125, 11, -103, -97, 115, -114], signature=[37, 46, 86, -25, -78, -82, -10, -76, -1, -82, 42, -76, -105, -37, -3, -55, -81, 40, -107, -7, -14, 126, -128, 66, -77, 51, 53, 12, 67, -88, -118, 19, -5, -60, -36, 46, 83, -91, -1, -87, -40, -22, 58, -1, -49, -50, 64, -24, -44, -98, 63, -22, -112, 76, -33, -54, -97, 94, 72, 82, -67, 102, 42, 2])], nativeScripts=[ScriptAll(type=all, scripts=[RequireTimeBefore(type=before, slot=62345848), ScriptPubkey(keyHash=bf75fe8a1272cbce74e89752104f5cd50be8380adb8e3cac73cde1ad, type=sig)])]), metadata=com.bloxbean.cardano.client.metadata.cbor.CBORMetadata@296810e8)`


Maybe I should do a policy sign on the original transaction before I pass it to nami like you suggested...

**Comment by satran004** [comment]
**Root cause:**
cardano-client-lib v0.1.5 is using cardano-serialization-lib (commit: af442ea) which doesn't have canonical ordering of assetname fix. But Nami is using a recent version of cardano-serialization-lib which has that fix. Because of that, ordering in multiasset list changes after policy signing and that causes the signature verification error.

**Solution:**
Use 0.2.0-preview1 or later if you want compatibility with Nami wallet for multiasset minting.

Here's the commit in cardano-serialization-lib for more info https://github.com/Emurgo/cardano-serialization-lib/commit/bab18feeb046d10f02c5616cd039872faab3600c

The canonical ordering will be supported in future release of client-lib to avoid any confusion.

**Comment by edridudi** [comment]
Resolved after moving to 0.2.0-preview1 version.
Issue Can be Closed.

Thank you very much @satran004 !
